text/0000-sigstore-integration.md

```diff
@@ -0,0 +1,387 @@
+* Make Sigstore signed crates mandatory.
+* Start marking legacy / unmaintained unsigned crates as unsafe
+* Add  metadata attributes to allow for identification of build source,  i.e. “certified” system such as GitHub actions. This could also take the form of using Sigstore’s organic countersigning capability.
+
+## Configuration changes to cargo
+
+By default, cargo will use the publicly available Sigstore instances together with GitHub as the identity provider. To override (if using a different registry than crates.io), the `$HOME/.cargo/sigstore.toml` file can be configured with the location of Sigstore and identity services:
+
+```
+fulcio = "https://fulcio.sigstore.dev"
+rekor = "https://rekor.sigstore.dev"
+issuer = "https://oauth2.sigstore.dev/auth"
+connector = "https://github.com/login/oauth"
+
+[registry]
+require-signed-crates =”yes/no”
+```
+
+## Signing
+
+Signing a crate happens when passing the `--sign` argument upon publishing a crate:
+
+```
+cargo publish --sign
+```
+
+Cargo will perform the necessary steps to sign your crate before publishing, and attach the relevant information to the crates.io request.
+
+Upon receiving the new crate metadata, crates.io will verify that the signature belongs to the crate owner.
+
+Signing an existing crate can be done using `cargo sign`:
+
+```
+cargo sign [<crate which you own>]
+```
+
+This will retrieve the crate from crates.io, and perform the same steps as for publish to generate the signature, but will attach the signature and certificate to the existing crate.
+
+
+## Operational Changes to Crates.io / Verification
+
+Crate signatures can be verified for any crate that is retrieved from `crates.io` that has signatures attached.
+
+To explicitly verify signatures for a crate, the `verify-signature` subcommand can be used:
+
+```
+cargo verify-signature [<crate>] [--skip-dependencies]
+```
+
+The command will fail if the crate (and dependencies) do not have verifiable signatures.
+
+Once the community has become familiar with signing and verification, signing and verification can be enabled by default for cargo commands.
+
+## Offline Verification
+
+Sigstore allows offline verification of crates by including additional metadata for the crate that can be stored just like other data for offline cargo usage.
+
+
+# Reference-Level Explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The crates.io and cargo types for new crates, will need to be modified to include the following metadata:
+
+* Signature - created based on the generated .crate file on publish
+* Identities - this must be one or more of permitted identities of the signer
+* Certificate - the public certificate that can be used to verify the crate
+* Bundle - allows for offline verification not connecting to the transparency log
+
+In the event that the crate is not being signed, these fields may be optional/null.
+## Identity, AuthN and AuthZ
+
+[identities]: #identities
+
+### Identity
```

Comment:
This section seems to do nothing other than explain what identities correspond to. The distinction of an NPE seems like something that should go in the motivation, alongside why such a distinction is important.